File: app/src/main/java/com/deepak/umber/data/repo/Netting.kt
package com.deepak.umber.data.repo

import kotlin.math.min

/** Money paid out to one counterparty, with the category it was filed under. */
data class MerchantDebit(val merchantNorm: String?, val category: String, val paise: Long)

/** Money received back from one counterparty. Income is excluded before this point. */
data class MerchantCredit(val merchantNorm: String, val paise: Long)

data class NettingResult(
    /** Everything paid out, before any settlement. */
    val grossPaise: Long,
    /** The part that came back from someone you had paid. */
    val reimbursedPaise: Long,
    /** What the spending actually cost: [grossPaise] − [reimbursedPaise]. */
    val netPaise: Long,
    /** Per-category totals, already net of reimbursements. */
    val netByCategory: Map<String, Long>,
)

/**
 * Nets reimbursements against spending, per counterparty.
 *
 * The problem: paying a friend ₹1,000 and being paid back ₹1,000 is not ₹1,000 of expense plus
 * ₹1,000 of income — it is ₹0. Counting both inflates spending *and* fakes income, and the
 * distortion compounds with every split bill.
 *
 * The rule is deliberately conservative: a credit only offsets spending when it comes from a
 * counterparty you actually paid **within the same window**, and it can never offset more than you
 * paid them. Money from someone you never paid is income, not a refund, so it is left alone.
 *
 * Two things keep salary out of it. The counterparty match does the heavy lifting — a paycheque
 * from an employer can only offset money paid *to* that employer, which nobody does. On top of
 * that, the caller drops credits the user has deliberately filed as income. Marking a paycheque as
 * a refund would silently erase real expenditure, which is the worst failure this calculation can
 * have, so it is guarded twice.
 *
 * Pure, so the arithmetic is unit-testable without a database.
 */
object Netting {

    fun apply(debits: List<MerchantDebit>, credits: List<MerchantCredit>): NettingResult {
        val gross = debits.sumOf { it.paise }

        val netByCategory = HashMap<String, Long>()
        for (debit in debits) {
            netByCategory[debit.category] = (netByCategory[debit.category] ?: 0L) + debit.paise
        }

        val creditByMerchant = credits
            .groupBy { it.merchantNorm }
            .mapValues { (_, rows) -> rows.sumOf { it.paise } }

        var reimbursed = 0L

        // A null merchant can't be matched to anyone, so those debits never offset.
        val debitsByMerchant = debits
            .filter { !it.merchantNorm.isNullOrBlank() }
            .groupBy { it.merchantNorm!! }

        for ((merchant, rows) in debitsByMerchant) {
            val credited = creditByMerchant[merchant] ?: continue
            val paidOut = rows.sumOf { it.paise }
            val offset = min(paidOut, credited)
            if (offset <= 0L) continue

            reimbursed += offset

            // Spread the offset across the categories this counterparty's spending was filed
            // under, in proportion to each. Integer division loses at most a few paise, so the
            // remainder is assigned to the largest row rather than silently dropped.
            var distributed = 0L
            val ordered = rows.sortedByDescending { it.paise }
            ordered.forEachIndexed { index, row ->
                val share = if (index == ordered.lastIndex) {
                    offset - distributed
                } else {
                    offset * row.paise / paidOut
                }
                distributed += share
                netByCategory[row.category] = (netByCategory[row.category] ?: 0L) - share
            }
        }

        return NettingResult(
            grossPaise = gross,
            reimbursedPaise = reimbursed,
            netPaise = gross - reimbursed,
            // A category can only be driven to zero, never negative: being paid back more than you
            // spent in a category is a data artefact, not negative spending.
            netByCategory = netByCategory.mapValues { (_, v) -> v.coerceAtLeast(0L) },
        )
    }
}

File: server/app/netting.py
"""Port of Netting.kt (app/src/main/java/com/deepak/umber/data/repo/Netting.kt).

Paying a friend and being paid back isn't spend + income, it's net zero. Offsets credits against
that merchant's debits, per merchant, up to the smaller of the two — never negative, never more
than was actually paid to them.

Same contract as the Kotlin original: the caller has already excluded credits filed as Income and
scoped both lists to one settlement window. This module only does the aggregation math.
"""

from collections import defaultdict
from dataclasses import dataclass


@dataclass
class MerchantDebit:
    merchant_norm: str | None
    category: str
    paise: int


@dataclass
class MerchantCredit:
    merchant_norm: str
    paise: int


@dataclass
class NettingResult:
    gross_paise: int
    reimbursed_paise: int
    net_paise: int
    net_by_category: dict[str, int]


def apply(debits: list[MerchantDebit], credits: list[MerchantCredit]) -> NettingResult:
    gross = sum(d.paise for d in debits)

    by_merchant: dict[str, dict[str, int]] = defaultdict(lambda: defaultdict(int))
    no_merchant_by_category: dict[str, int] = defaultdict(int)
    for d in debits:
        if d.merchant_norm:
            by_merchant[d.merchant_norm][d.category] += d.paise
        else:
            no_merchant_by_category[d.category] += d.paise

    credit_by_merchant: dict[str, int] = defaultdict(int)
    for c in credits:
        credit_by_merchant[c.merchant_norm] += c.paise

    net_by_category: dict[str, int] = defaultdict(int)
    reimbursed = 0

    for merchant, categories in by_merchant.items():
        merchant_total = sum(categories.values())
        credit_available = credit_by_merchant.get(merchant, 0)
        offset = min(merchant_total, credit_available)

        if offset <= 0:
            for category, amount in categories.items():
                net_by_category[category] += amount
            continue

        reimbursed += offset
        items = list(categories.items())
        remaining_offset = offset
        for i, (category, amount) in enumerate(items):
            if i == len(items) - 1:
                category_offset = remaining_offset
            else:
                category_offset = round(offset * amount / merchant_total)
                remaining_offset -= category_offset
            net_by_category[category] += amount - category_offset

    for category, amount in no_merchant_by_category.items():
        net_by_category[category] += amount

    net = gross - reimbursed
    return NettingResult(
        gross_paise=gross,
        reimbursed_paise=reimbursed,
        net_paise=net,
        net_by_category=dict(net_by_category),
    )
